src/backends/plonky2/primitives/ec/gates/curve.rs

```diff
@@ -25,30 +49,848 @@ use crate::backends::plonky2::primitives::ec::{
+            array::from_fn(|i| builder.add_const_extension(zero, g_u[i])),
+        ];
+
+        let double_constraint =
+            |builder: &mut CircuitBuilder<GoldilocksField, D>, i: usize, j: usize| {
+                let x1 = array::from_fn(|k| vars.local_wires[i + k]);
+                let u1 = array::from_fn(|k| vars.local_wires[i + 5 + k]);
+                let [x, z, u, t] = homog_ec_target_add(builder, &x1, &u1, &x1, &u1);
+
+                let mut new_constraints = Vec::<ExtensionTarget<_>>::with_capacity(10);
+                let x2 = array::from_fn(|k| vars.local_wires[j + k]);
+                let u2 = array::from_fn(|k| vars.local_wires[j + 5 + k]);
+
+                let [expected_x, expected_u] = [
+                    nnf_mul_ext::<D, 5, Nnf>(builder, &x2, &z),
+                    nnf_mul_ext::<D, 5, Nnf>(builder, &u2, &t),
+                ];
+                let first_constraints = nnf_ext_target_sub::<D, 5, Nnf>(builder, &expected_x, &x);
+                let second_constraints = nnf_ext_target_sub::<D, 5, Nnf>(builder, &expected_u, &u);
+
+                new_constraints.extend(&first_constraints);
+                new_constraints.extend(&second_constraints);
+                new_constraints
+            };
+
+        let select_and_add_constraint = |builder: &mut CircuitBuilder<GoldilocksField, D>,
+                                         i: usize,
+                                         j: usize,
+                                         selector_index: usize,
+                                         add_y: bool| {
+            let zero = builder.zero_extension();
+
+            let one = builder.one_extension();
+            let one_full = [one, zero, zero, zero, zero];
+
+            let sel = vars.local_wires[selector_index];
+            let sel_full = [sel, zero, zero, zero, zero];
+
+            let x1 = array::from_fn(|k| vars.local_wires[i + k]);
+            let u1 = array::from_fn(|k| vars.local_wires[i + 5 + k]);
+
+            let (x2, u2) = if add_y {
+                // (using hardcoded location of y)
+                (
+                    array::from_fn(|k| vars.local_wires[10 + k]),
+                    array::from_fn(|k| vars.local_wires[15 + k]),
+                )
+            } else {
+                (g_x_ext_target, g_u_ext_target)
+            };
+
+            let [x, z, u, t] = homog_ec_target_add(builder, &x1, &u1, &x2, &u2);
+
+            let mut new_constraints = Vec::<ExtensionTarget<_>>::with_capacity(10);
+            let x3 = array::from_fn(|k| vars.local_wires[j + k]);
+            let u3 = array::from_fn(|k| vars.local_wires[j + 5 + k]);
+
+            let sel_one_diff = nnf_ext_target_sub::<D, 5, Nnf>(builder, &sel_full, &one_full);
+            let first_constraints = {
+                let term1 = nnf_mul_ext::<D, 5, Nnf>(builder, &x3, &z);
+                let term2 = nnf_mul_ext::<D, 5, Nnf>(builder, &sel_full, &x);
```

Comment:
Since `sel_full` is a scalar embedded in the extension, I guess this mul could be implemented in a more efficient way?
Is it just this?
```rust
[builder.mul_extension(sel, x[0]), builder.mul_extension(sel, x[1]), builder.mul_extension(sel, x[2]), builder.mul_extension(sel, x[3]), builder.mul_extension(sel, x[4])]
```